libappstream-builder/plugins/asb-plugin-sugar-activity.c

```diff
@@ -0,0 +1,286 @@
+/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
+ *
+ * Copyright (C) 2016 Sam Parkinson <sam@sam.today>
+ *
+ * Licensed under the GNU Lesser General Public License Version 2.1
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <config.h>
+#include <string.h>
+#include <fnmatch.h>
+#include <gdk-pixbuf/gdk-pixbuf.h>
+
+#include <asb-plugin.h>
+
+#define __APPSTREAM_GLIB_PRIVATE_H
+#include <as-utils-private.h>
+#include <as-app-private.h>
+
+/**
+ * asb_plugin_get_name:
+ */
+const gchar *
+asb_plugin_get_name (void)
+{
+	return "sugar-activity";
+}
+
+#define _SUGAR_ACTIVITY_GLOB "/usr/share/sugar/activities/*/activity/activity.info"
+#define _SUGAR_ACTIVITY_ICON_GLOB "/usr/share/sugar/activities/*/activity/*"
+#define _SUGAR_ACTIVITY_LINFO_GLOB "/usr/share/sugar/activities/*/locale/*/activity.linfo"
+
+/**
+ * asb_plugin_add_globs:
+ */
+void
+asb_plugin_add_globs (AsbPlugin *plugin, GPtrArray *globs)
+{
+	// the _SUGAR_ACTIVITY_GLOB is a subset of the icon glob
+	asb_plugin_add_glob (globs, _SUGAR_ACTIVITY_ICON_GLOB);
+	asb_plugin_add_glob (globs, _SUGAR_ACTIVITY_LINFO_GLOB);
+}
+
+/**
+ * _asb_plugin_is_actinfo
+ */
+static gboolean
+_asb_plugin_is_actinfo (const gchar *filename)
+{
+	if (asb_plugin_match_glob (_SUGAR_ACTIVITY_GLOB, filename))
+		return TRUE;
+	return FALSE;
+}
+
+/**
+ * _asb_plugin_is_linfo
+ */
+static gboolean
+_asb_plugin_is_linfo (const gchar *filename)
+{
+	if (asb_plugin_match_glob (_SUGAR_ACTIVITY_LINFO_GLOB, filename))
+		return TRUE;
+	return FALSE;
+}
+
+/**
+ * asb_plugin_check_filename:
+ */
+gboolean
+asb_plugin_check_filename (AsbPlugin *plugin, const gchar *filename)
+{
+	return _asb_plugin_is_linfo (filename) ||
+	       _asb_plugin_is_actinfo (filename);
+}
+
+static gboolean
+_asb_plugin_load_sugar_icon (const gchar *path,
+			     AsIcon *ic) {
+	gsize len;
+	gchar *data = NULL;
+	GBytes *bytes = NULL;
+
+	if (!g_file_get_contents (path, &data, &len, NULL))
+		return FALSE;
+	printf ("ICON (got contents) = %s\n", path);
+
+	as_icon_set_kind (ic, AS_ICON_KIND_EMBEDDED);
+	bytes = g_bytes_new_take (data, len);
+	as_icon_set_data (ic, bytes);
+	as_icon_set_width (ic, 75);
```

Comment:
An application can not ship multiple different icons and it is not allowed to ship an SVG icon either, for the reasons @hughsie explained.
But you don't need that, actually, since appstream-builder will render the SVG to PNG icons (I suppose ^^).
I also don't see the use case for having apps have multiple different icons, if anything, it adds more confusion to the users. The only case where this is probably useful is for monochromatic low-res icons for accessibility reasons.

Btw, on a more general note, I don't think this plugin is great, since it is exclusive for appstream-builder, meaning only stuff using that tool to generate AppStream metadata will benefit from it. Additionally, converting formats usually means loss of information, since AppStream metainfo files usually can contain much more data than the source format that you convert into them. Having a "magic" conversion removes the need for writing metainfo files for the respective upstream project, making it ship less metadata in comparison.
It also prevents other tools which handle metainfo files from picking them up when a sugar app is installed.
So IMHO this is a nice workaround to show those apps quickly, but you should really write metainfo files for these apps.
Cheers,
    Matthias